impl-tinkerpop-parent/tinkergraph-provider/src/main/java/org/hawkular/inventory/impl/tinkerpop/provider/TinkerGraphProvider.java

```diff
@@ -39,6 +45,47 @@ public void ensureIndices(WrappedTinkerGraph graph, IndexSpec... indexSpecs) {
         //don't bother with this for a demo graph
     }
 
+    @Override
+    public InventoryBackend.Transaction startTransaction(WrappedTinkerGraph graph, boolean mutating) {
+        ReentrantReadWriteLock lock;
+        synchronized (transactionLocks) {
+            lock = transactionLocks.get(graph);
+            if (lock == null) {
+                lock = new ReentrantReadWriteLock();
+                transactionLocks.put(graph, lock);
+            }
+        }
+
+        if (mutating) {
+            lock.writeLock().lock();
+        } else {
+            lock.readLock().lock();
+        }
+
+        return new SimulatedSerializedTransaction(mutating, lock);
+    }
+
+    @Override
+    public void commit(WrappedTinkerGraph graph, InventoryBackend.Transaction t) {
+        graph.commit();
+        unlock(t);
```

Comment:
If you look at the implementation of `WrappedTinkergraph#commit()`, you'll notice it is a no-op. But your solution is technically more correct. I'll update the code.

The assumption with the locks is that the user (the base impl) does it right. I would not go further with Tinkergraph impl - timeouts would complicate the matters greatly because they could corrupt the database in the same way it could be corrupted prior to HAWKULAR-255.

Titan and any other implementation has the opportunity to do this differently and in fact the default implementations of the methods on the interface actually use no locks at all because they assume the underlying graph is truly transactional (which TinkerGraph is not).